assignments/Session_02/tests.py

```diff
@@ -0,0 +1,450 @@
+                "mimetype {0} not in response for {1}".format(
+                    expected_mimetype, file_uri
+                )
+            )
+            self.assertTrue(
+                expected_body in actual,
+                "body of {0} not in response for {1}".format(
+                    file_path, file_uri
+                )
+            )
+
+    def test_webroot_image_uris(self):
+        """verify that image uris are properly served
+
+        requires using a client that does not attempt to decode the response
+        body
+        """
+        message_tmpl = CRLF.join(['GET {0} HTTP/1.1', 'Host: example.com', ''])
+        root = pathlib.Path("webroot")
+        images_path = root / 'images'
+        for file_path in images_path.iterdir():
+            # set up expectations for this file
+            if file_path.is_dir():
+                continue
+            expected_body = file_path.read_bytes()
+            expected_mimetype = mimetypes.types_map[
+                os.path.splitext(str(file_path))[1]
+            ]
+            file_uri = str(file_path)[len(str(root)):]
+            message = message_tmpl.format(file_uri)
+            actual = self.send_message(message, use_bytes=True)
+            self.assertTrue(
+                b"200 OK" in actual,
+                "request for {0} did not result in OK".format(
+                    file_uri
+                )
+            )
+            self.assertTrue(
+                expected_mimetype.encode('utf8') in actual,
+                "mimetype {0} not in response for {1}".format(
+                    expected_mimetype, file_uri
+                )
+            )
+            self.assertTrue(
+                expected_body in actual,
+                "body of {0} not in response for {1}".format(
+                    file_path, file_uri
+                )
+            )
+
+    def test_missing_resource(self):
+        message = CRLF.join(
+            ['GET /missing.html HTTP/1.1', 'Host: example.com', '']
+        )
+        expected = '404 Not Found'
+        actual = self.send_message(message)
+        self.assertTrue(
+            expected in actual, '"{0}" not in "{1}"'.format(expected, actual)
+        )
+
+
+if __name__ == '__main__':
+    unittest.main()
+
+#Debugging stuff follows
+"""
+if __name__ == '__main__':
+    suite = unittest.TestLoader().loadTestsFromTestCase(HTTPServerFunctionalTestCase)
+    unittest.TextTestRunner(verbosity=2).run(suite)
+"""
+
+"""
+if __name__ == '__main__':
+    suite = unittest.TestLoader().loadTestsFromTestCase(HTTPServerFunctionalTestCase)
+    unittest.TextTestRunner(verbosity=2).run(suite)
+"""
+###python tests.py ResolveURITestCase.test_file_resource
+###python tests.py HTTPServerFunctionalTestCase.test_get_request
+####import pdb; pdb.set_trace()    ###import debugger
+#print('!!!!! function_under_test mimetype_one =', mimetype, '\n', file=sys.stderr)  #debug - remove !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Comment:
Try to remove dead or unused code from your work before submitting.  It's a good habit to get into.